docs/examples/move/nft_marketplace/sources/rental_extension.move

```diff
@@ -0,0 +1,417 @@
+
+    // rules imports
+    use kiosk::kiosk_lock_rule::Rule as LockRule;
+
+    // === Errors ===
+    const EExtensionNotInstalled: u64 = 0;
+    const ENotOwner: u64 = 1;
+    const ENotEnoughCoins: u64 = 2;
+    const EInvalidKiosk: u64 = 3;
+    const ERentingPeriodNotOver: u64 = 4;
+    const EObjectNotExist: u64 = 5;
+    const ETotalPriceOverflow: u64 = 6;
+
+    // === Constants ===
+    const PERMISSIONS: u128 = 11;
+    const SECONDS_IN_A_DAY: u64 = 86400;
+    const MAX_BASIS_POINTS: u16 = 10_000;
+    const MAX_VALUE_U64: u64 = 0xff_ff_ff_ff__ff_ff_ff_ff;
+
+    // === Structs ===
+
+    /// Extension Key for Kiosk Marketplace extension.
+    public struct Rental has drop {}
+
+    /// Struct representing a rented item.
+    /// Used as a key for the Rentable that's placed in the Extension's Bag.
+    public struct Rented has store, copy, drop { id: ID }
+
+    /// Struct representing a listed item.
+    /// Used as a key for the Rentable that's placed in the Extension's Bag.
+    public struct Listed has store, copy, drop { id: ID }
+
+    /// Promise struct for borrowing by value.
+    public struct Promise {
+        item: Rented,
+        duration: u64,
+        start_date: u64,
+        price_per_day: u64,
+        renter_kiosk: ID,
+        borrower_kiosk: ID,
+    }
+
+    /// A wrapper object that holds an asset that is being rented.
+    /// Contains information relevant to the rental period, cost and renter.
+    public struct Rentable<T: key + store> has store {
+        object: T,
+        /// Total amount of time offered for renting in days.
+        duration: u64,
+        /// Initially undefined, is updated once someone rents it.
+        start_date: Option<u64>,
+        price_per_day: u64,
+        /// The kiosk id that the object was taken from.
+        kiosk_id: ID,
+    }
+
+    /// A shared object that should be minted by every creator.
+    /// Defines the royalties the creator will receive from each rent invocation.
+    public struct RentalPolicy<phantom T> has key, store {
+        id: UID,
+        balance: Balance<IOTA>,
+        /// Note: Move does not support float numbers.
+        ///
+        /// If you need to represent a float, you need to determine the desired
+        /// precision and use a larger integer representation.
+        ///
+        /// For example, fpercentages can be represented using basis points:
+        /// 10000 basis points represent 100% and 100 basis points represent 1%.
+        amount_bp: u64
+    }
+
+    /// A shared object that should be minted by every creator.
+    /// Even for creators that do not wish to enforce royalties. Provides authorized access to an
+    /// empty TransferPolicy.
+    public struct ProtectedTP<phantom T> has key, store {
+        id: UID,
+        transfer_policy: TransferPolicy<T>,
+        policy_cap: TransferPolicyCap<T>
+    }
+
+    // === Public Functions ===
+
+    /// Enables someone to install the Marketplace extension in their Kiosk.
```

Comment:
check comment here and later (marketplace?)

Reply:
Done